Fix multiple index field support

jannon opened the pull request:
Okay, this ended up being more involved than I originally thought, but here's a PR with enhanced support for using multiple indexes.  I'll describe the following major pieces below:
* Automatic Multiple Index Field Handling
* Ability to Use Multiple Serializers
* Serializer Mixin
* Miscellaneous

## Automatic Multiple Index Field Handling

The original problem that needed to be solved was that if you used multiple indexes that had overlapping field names, only one field would be used.  If they happened to be different field types, that would be a big problem.

The code in this PR internally modifies the field names to associate them with a specific index.  Let's say we have the following setup:

```python
class Index1(indexes.searchIndex, indexes.Indexable):
    text = indexes.CharField(document=True, use_template=True)
    my_field = indexes.CharField()
    ...

class Index2(indexes.searchIndex, indexes.Indexable):
    text = indexes.CharField(document=True, use_template=True)
    my_field = indexes.IntegerField()
    ...

class MySerializer(HaystackSerializer):
    class Meta:
        index_classes = [Index1, Index2]
```

Internally, the fields will be given the following names:
* `_Index1__text`
* `_Index2__text`
* `_Index1__my_field`
* `_Index2__my_field`

Thus, no overlap issues occur.  The field names are converted back during serialization so the handling is transparent to the user.  To make this work, the classes in the `_field_mapping` had to be replaced with proxy versions that handle the renaming

## Ability to Use Multiple Serializers

As an alternative to the automatic handling above, a user may choose to specify a dictionary of Index:Serializer mappings to user a separate serializer of his or her choosing for each index.  Let's say we also have:

```python
class Serializer1(serializers.Serializer):
    ...

class Serializer2(serializers.Serializer):
    ...

class MySerializer2(HaystackSerializer):
    class Meta:
        serializers = {
            Index1: Serializer1,
            Index2: Serializer2
        }
```

Each result will be serialized by the appropriate serializer.  `serializers` takes precedent if `index_classes` is also specified.

## Serializer Mixin

A `HaystackSerializerMixin` class is provided as convenience to easily reuse existing model serializers.  This can be desireable if you want to return the same data formats from your search APIs that you do from your other APIs and have complex formats that you don't want to recreate with search indexes.  And if you don't mind the performance hit.  We might have something like:

```python
class Model1Serializer(serializers.ModelSerializer):
    ...

class Model2Serializer(serializers.ModelSerializer):
    ...

class Model1SearchSerializer(HaystackSerializerMixin, Model1Serializer):
    pass

class Model2SearchSerializer(HaystackSerializerMixin, Model2Serializer):
    pass

class MySerializer3(HaystackSerializer):
    class Meta:
        serializers = {
            Index1: Model1SearchSerializer,
            Index2: Model2SearchSerializer
        }
```

## Miscellaneous

This PR comes with updated tests and documentation, as well, including a migration and fixture for a third mock model (MockPet) in the test app.  

Moving forward, I'm considering adding an option to have the serializer automatically inject a 'type' field or the like, to more readily facilitate post-processing of the results list (e.g. separating out results by type), but for the time being users can achieve this themselves by adding something to their serializers or indexes

rhblind commented:
Hi!
Wow, thanks a bunch, awesome work!
I need to take some time to look at it, but it looks really good. I'll comment a little inline on the changes if I have any questions.

Time to make a Contributers list ;)

rhblind commented on drf_haystack/serializers.py:
```diff
@@ -131,24 +138,69 @@ def get_fields(self):
+        Converts in index model class to a name suitable for use as a field name prefix. A user
+        may optionally specify custom aliases via an 'index_aliases' attribute on the Meta class
+        """
+        cls_name = str(index_cls).replace("<class '", "").replace("'>", "")
```
For me it seems a bit safer to use `cls_name = index_cls.__class__.__name__` in order to get the class name. The method you have used is based on the class `__repr__` (or `__str__` can't remember), and feels a bit unsafe to me.

jannon replied:
Absolutely, definitely better.  I'll make that change

rhblind commented on drf_haystack/serializers.py:
```diff
@@ -131,24 +138,69 @@ def get_fields(self):
+            raise ImproperlyConfigured("Could not find serializer for %s in mapping" % index)
+        return serializer_class(context=self._context).to_representation(instance)
+
+    def default_to_representation(self, instance):
```
Wouldn't it be more intuitive to let the `default_to_representation` code live in the default `to_representation` method, and the custom code to live in a custom method? Or even better if they could co-exist in the same method.
I had to read this twice to understand the flow here, and in my head; since the "custom code" is just a few lines, it could easily live in the same method.

rhblind replied:
Also, we probably want to support `highlighted` fields in the serializer mappings.

jannon replied:
I think my intention was a modular composition that would allow downstream developers the flexibility to modify just the parts they wanted to when subclassing.  But now that I think about it, a cleaner solution for that might be something like (pseudo-code):

```python
def to_representation(self, instance):
    if we_have_serializer_mapping:
        return self.multi_serializer_representation(instance)
    return self.regular_representation(instance)
```

As one who's used many open-source projects and often been frustrated by inflexibility, I'd really like to separate pieces of functionality like this so another developer could just change the mapping behaviour or just the regular behavior if they wanted to.

If you have suggestions for a better name than `regular_representation`...

As for the `highlighted` fields...yeah, I haven't actually ever even looked at the highlighter code.  I'm just scanning `HighlighterMixin` now...okay, I'll have to take a look at it.

rhblind replied:
Yeah, I agree modularity is a good thing, we just have to make sure we don't confuse people further by adding too many (new) steps in an already familiar chain.
Most people using drf already expects the `to_representation` method to do a certain thing, so we need to make our intent clear if we're heading down a too modular path.

So I thinks something like this would be good

    def to_representation(self, instance):
        if we_have_serializer_mapping:
            return self.mulit_serializer_representation(instance)   
        ...
        default code here
        ....

But, I'm not going to force this through, so if you think the modular approach is better, go for it ;)

As for the `highlighted` fields, I believe it's pretty straight forward.

rhblind commented on docs/advanced_usage.rst:
```diff
@@ -354,3 +354,183 @@ restrictions on the views. This can be overridden on a per-view basis as you wou
+
+.. note::
+
+    The ``AggregateSearchViewSet class above omits the optional ``index_models`` attribute.  This way results from all the
```
Not to be picky, but a minor typo here. Seems you forgot to close the first quotes? Should probably be ``AggregateSearchView`` class above...

jannon replied:
Ah, yep...on that note...nevermind.  I was about to ask if there an easy way for me to render the docs to test, but I just saw the Makefile in docs

jannon commented:
Okay, let's see.  I swapped the branch logic around in `to_representation`, used the more robust class name retrieval, and fixed some docs typos.  

The highlighted field is now being added regardless of serializer route.  I was confused at first because I kind of feel that if the developer is going the multi serializer route, they should be responsible for handling that in their serializer.  Indeed, a developer may not want extra fields added.  They could have some non-robust deserializer that explodes in the presence of the unexpected field.  I don't feel strongly about it, but I thought I'd bring it up

Also, I added support for query field negation because I needed it right away.  So specifying a query like so

```
https://whatever.com/api/v1/search?text__not=foo
```

will return all results without 'foo'.  The negation keyword (default: 'not') is configurable in the settings.

rhblind commented:
Goodie stuff, Jannon!
I'm not home right now, but I'll take a look at it tomorrow, and merge it.
I'll probably make it a point release to pypi as well.

Thanks for taking time to help me with the library, I really appreciate it
:)
On Aug 22, 2015 5:17 AM, "Jannon Frank" <notifications@github.com> wrote:

> Okay, let's see. I swapped the branch logic around in to_representation,
> used the more robust class name retrieval, and fixed some docs typos.
>
> The highlighted field is now being added regardless of serializer route. I
> was confused at first because I kind of feel that if the developer is going
> the multi serializer route, they should be responsible for handling that in
> their serializer. Indeed, a developer may not want extra fields added. They
> could have some non-robust deserializer that explodes in the presence of
> the unexpected field. I don't feel strongly about it, but I thought I'd
> bring it up
>
> Also, I added support for query field negation because I needed it right
> away. So specifying a query like so
>
> https://whatever.com/api/v1/search?text__not=foo
>
> will return all results without 'foo'. The negation keyword (default:
> 'not') is configurable in the settings.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inonit/drf-haystack/pull/23#issuecomment-133630315>.
>

rhblind commented:
Yeah, about the serialization of highlighted filds, the extra field is only added if using the HighlightFilter backend, or the mixin. I agree that maybe the developer should be responsible for serialization of those fields, but it turned out to be not too intuitive to look for the `highlighted` attribute, so I just put it in there. So the extra field is only added if the developer actually want highlighting..

This is of course open for discussion if this seems a bit intrusive, but we can make another issue for that.

rhblind commented:
Merged and uploaded v1.5.2 to pypi.
Thanks alot, Jannon!

jannon commented:
Great!  Yeah, thanks for creating the project
